### PR TITLE
doc: add tables with supported features to nrf_security

### PR DESCRIPTION
**Pull request by b-gent**
Update the nrf_security RST file with tables listing the features
supported by different cryptographic algorithms.

Signed-off-by: Bartosz Gentkowski <bartosz.gentkowski@nordicsemi.no>

This is the current output, some information is missing and needs to be provided, marked as TBD.
[Backend configurations and supported features — nrfxlib 1.4.99 documentation.pdf](https://github.com/nrfconnect/sdk-nrfxlib/files/5932417/Backend.configurations.and.supported.features.nrfxlib.1.4.99.documentation.pdf)

I'm still figuring out if line breaks are possible within a cell.

### CODE REVIEW

File: nrf_security/doc/backend_config.rst [hunk 1]
```diff
@@ -69,11 +69,24 @@ AES core support can be enabled by setting setting the :option:`CONFIG_MBEDTLS_A
...
 
+Feature support
+===============
+
++-------------+-------------------+-------------+
+| Cipher mode | Backend           | Key size    |
++=============+===================+=============+
+| ECB         | cc3xx             | 128-bit key |
+|             +-------------------+-------------+
+|             | nrf_oberon        | TBD         |
```
Comment (frkv) [inline nrf_security/doc/backend_config.rst:84]: nrf_oberon supports 128-bit, 192-bit and 256-bit

File: nrf_security/doc/backend_config.rst [hunk 2]
```diff
@@ -102,12 +115,53 @@ AES cipher modes can be enabled by setting one or more of the following Kconfig
...
 
+Feature support
+===============
+
++-------------+-------------------+-------------+-----------------------+
+| Cipher mode | Backend           | Key size    | Note                  |
++=============+===================+=============+=======================+
+| CTR         | cc3xx             | 128-bit key |                       |
+|             +-------------------+-------------+-----------------------+
+|             | nrf_oberon        | TBD         |                       |
```
Comment (frkv) [inline nrf_security/doc/backend_config.rst:130]: All supported, see previous comment

File: nrf_security/doc/backend_config.rst [hunk 3]
```diff
@@ -102,12 +115,53 @@ AES cipher modes can be enabled by setting one or more of the following Kconfig
...
+|             +-------------------+-------------+-----------------------+
+|             | Original mbed TLS | 128-bit key |                       |
+|             |                   | 192-bit key |                       |
+|             |                   | 256-bit key |                       |
++-------------+-------------------+-------------+-----------------------+
+| CBC         | cc3xx             | 128-bit key |                       |
+|             +-------------------+-------------+-----------------------+
+|             | nrf_oberon        | TBD         |                       |
```
Comment (frkv) [inline nrf_security/doc/backend_config.rst:138]: All supported, see previous comment

File: nrf_security/doc/backend_config.rst [hunk 4]
```diff
@@ -102,12 +115,53 @@ AES cipher modes can be enabled by setting one or more of the following Kconfig
...
+|             +-------------------+-------------+-----------------------+
+|             | Original mbed TLS | 128-bit key |                       |
+|             |                   | 192-bit key |                       |
+|             |                   | 256-bit key |                       |
++-------------+-------------------+-------------+-----------------------+
+| CFB         | cc3xx             | N/A         | Backend not supported |
+|             +-------------------+-------------+-----------------------+
+|             | nrf_oberon        | TBD         |                       |
```
Comment (frkv) [inline nrf_security/doc/backend_config.rst:146]: All supported, see previous comment

File: nrf_security/doc/backend_config.rst [hunk 5]
```diff
@@ -102,12 +115,53 @@ AES cipher modes can be enabled by setting one or more of the following Kconfig
...
+|             +-------------------+-------------+-----------------------+
+|             | Original mbed TLS | 128-bit key |                       |
+|             |                   | 192-bit key |                       |
+|             |                   | 256-bit key |                       |
++-------------+-------------------+-------------+-----------------------+
+| OFB         | cc3xx             | N/A         | Backend not supported |
+|             +-------------------+-------------+-----------------------+
+|             | nrf_oberon        | TBD         |                       |
```
Comment (frkv) [inline nrf_security/doc/backend_config.rst:154]: All supported, see previous comment

File: nrf_security/doc/backend_config.rst [hunk 6]
```diff
@@ -102,12 +115,53 @@ AES cipher modes can be enabled by setting one or more of the following Kconfig
...
+|             +-------------------+-------------+-----------------------+
+|             | Original mbed TLS | TBD         |                       |
++-------------+-------------------+-------------+-----------------------+
+| XTS         | cc3xx             | N/A         | Backend not supported |
+|             +-------------------+-------------+-----------------------+
+|             | nrf_oberon        | TBD         |                       |
```
Comment (frkv) [inline nrf_security/doc/backend_config.rst:160]: All supported, see previous comment

File: nrf_security/doc/backend_config.rst [hunk 7]
```diff
@@ -144,6 +198,22 @@ CMAC can be enabled by setting the :option:`CONFIG_MBEDTLS_CMAC_C` Kconfig varia
...
 
+Feature support
+===============
+
++-------------+-------------------+-----------------------+
+| Algorithm   | Backend           | Key size              |
++=============+===================+=======================+
+| CMAC        | cc3xx             | 128-bit key           |
+|             +-------------------+-----------------------+
+|             | nrf_oberon        | TBD                   |
```
Comment (frkv) [inline nrf_security/doc/backend_config.rst:209]: All supported, see previous comment

**Comment by b-gent** [comment]
Updated with values for nrf_oberon
[Backend configurations and supported features 2.pdf](https://github.com/nrfconnect/sdk-nrfxlib/files/5940277/Backend.configurations.and.supported.features.2.pdf)

File: nrf_security/doc/backend_config.rst [hunk 8]
```diff
@@ -102,12 +117,63 @@ AES cipher modes can be enabled by setting one or more of the following Kconfig
...
+|             |                   | 192-bit key |                       |
+|             |                   | 256-bit key |                       |
++-------------+-------------------+-------------+-----------------------+
+| OFB         | cc3xx             | N/A         | Backend not supported |
+|             +-------------------+-------------+-----------------------+
+|             | nrf_oberon        | 128-bit key |                       |
+|             |                   | 192-bit key |                       |
+|             |                   | 256-bit key |                       |
+|             +-------------------+-------------+-----------------------+
+|             | Original mbed TLS | TBD         |                       |
```
Comment (b-gent) [inline nrf_security/doc/backend_config.rst:166]: all key sizes

File: nrf_security/doc/backend_config.rst [hunk 9]
```diff
@@ -102,12 +117,63 @@ AES cipher modes can be enabled by setting one or more of the following Kconfig
...
++-------------+-------------------+-------------+-----------------------+
+| XTS         | cc3xx             | N/A         | Backend not supported |
+|             +-------------------+-------------+-----------------------+
+|             | nrf_oberon        | 128-bit key |                       |
+|             |                   | 192-bit key |                       |
+|             |                   | 256-bit key |                       |
+|             +-------------------+-------------+-----------------------+
+|             | Original mbed TLS | TBD         |                       |
```
Comment (b-gent) [inline nrf_security/doc/backend_config.rst:174]: all

File: nrf_security/doc/backend_config.rst [hunk 10]
```diff
@@ -267,12 +349,53 @@ ChaCha-Poly can be enabled by setting the :option:`CONFIG_MBEDTLS_CHACHAPOLY_C`
...
+|              |                   | 256-bit key |                                                                      |
++--------------+-------------------+-------------+----------------------------------------------------------------------+
+| ChaCha20     | cc3xx             | TBD         |                                                                      |
+|              +-------------------+-------------+----------------------------------------------------------------------+
+|              | nrf_oberon        | TBD         |                                                                      |
+|              +-------------------+-------------+----------------------------------------------------------------------+
+|              | Original mbed TLS | TBD         |                                                                      |
++--------------+-------------------+-------------+----------------------------------------------------------------------+
+| Poly1305     | cc3xx             | TBD         |                                                                      |
```
Comment (b-gent) [inline nrf_security/doc/backend_config.rst:381]: 256-bit for all , Poly, Chacha, and Chacha-Poly

File: nrf_security/doc/backend_config.rst [hunk 11]
```diff
@@ -301,9 +422,18 @@ DHM can be enabled by setting the :option:`CONFIG_MBEDTLS_DHM_C` Kconfig variabl
...
+===============
 
++-----------+-------------------+----------------------+
+| Algorithm | Backend           | Key size             |
++===========+===================+======================+
+| DHM       | cc3xx             | Limited to 2048 bits |
+|           +-------------------+----------------------+
+|           | nrf_oberon        | TBD                  |
+|           +-------------------+----------------------+
+|           | Original mbed TLS | TBD                  |
```
Comment (b-gent) [inline nrf_security/doc/backend_config.rst:435]: No limitation

File: nrf_security/doc/backend_config.rst [hunk 12]
```diff
@@ -301,9 +422,18 @@ DHM can be enabled by setting the :option:`CONFIG_MBEDTLS_DHM_C` Kconfig variabl
...
-   * :ref:`nrf_security_backends_cc3xx`  is limited to key size of <= 2048 bits.
+Feature support
+===============
 
++-----------+-------------------+----------------------+
+| Algorithm | Backend           | Key size             |
++===========+===================+======================+
+| DHM       | cc3xx             | Limited to 2048 bits |
+|           +-------------------+----------------------+
+|           | nrf_oberon        | TBD                  |
```
Comment (b-gent) [inline nrf_security/doc/backend_config.rst:433]: not supported

File: nrf_security/doc/backend_config.rst [hunk 13]
```diff
@@ -336,7 +466,58 @@ ECC core support can be enabled by setting the :option:`CONFIG_MBEDTLS_ECP_C` Kc
 
 .. note::
    * For features provided with :ref:`Choice<nrf_security_backend_config_multiple>` support, the enabled backend that is first in order is selected by default.
-   * The :ref:`nrf_oberon_readme` only supports ECC curve secp256r1.
+
+Feature support
+===============
+
++-----------+-------------------+----------+-----------------------------+
+| Algorithm | Backend           | Key size | Curve type                  |
```
Comment (b-gent) [inline nrf_security/doc/backend_config.rst:474]: skip the key size column

File: nrf_security/doc/backend_config.rst [hunk 14]
```diff
@@ -336,7 +466,58 @@ ECC core support can be enabled by setting the :option:`CONFIG_MBEDTLS_ECP_C` Kc
...
+|           |                   |          | secp160k1                   |
+|           |                   |          | secp192k1                   |
+|           |                   |          | secp224k1                   |
+|           |                   |          | secp256k1                   |
+|           |                   |          |                             |
+|           |                   |          | Curve25519 group:           |
+|           |                   |          | Curve25519                  |
+|           +-------------------+----------+-----------------------------+
+|           | nrf_oberon        | TBD      | NIST group:                 |
+|           |                   |          | secp256r1                   |
```
Comment (b-gent) [inline nrf_security/doc/backend_config.rst:495]: add secp224r1

File: nrf_security/doc/backend_config.rst [hunk 15]
```diff
@@ -354,8 +535,59 @@ Elliptic Curve Diffie-Hellman (ECDH) support can be enabled by setting the :opti
    * ECDH support depends upon `ECC Configurations`_ being enabled.
    * The :ref:`nrf_cc3xx_mbedcrypto_readme` does not integrate on ECP layer.
      Only the top-level APIs for ECDH are replaced.
-   * The :ref:`nrf_oberon_readme` only supports ECC curve secp256r1.
 
+Feature support
+===============
+
++-----------+-------------------+----------+-------------------+
+| Algorithm | Backend           | Key size | Curve type        |
```
Comment (b-gent) [inline nrf_security/doc/backend_config.rst:543]: remove key size

File: nrf_security/doc/backend_config.rst [hunk 16]
```diff
@@ -354,8 +535,59 @@ Elliptic Curve Diffie-Hellman (ECDH) support can be enabled by setting the :opti
...
+|           |                   |          | secp160k1         |
+|           |                   |          | secp192k1         |
+|           |                   |          | secp224k1         |
+|           |                   |          | secp256k1         |
+|           |                   |          |                   |
+|           |                   |          | Curve25519 group: |
+|           |                   |          | Curve25519        |
+|           +-------------------+----------+-------------------+
+|           | nrf_oberon        | TBD      | NIST group:       |
+|           |                   |          | secp256r1         |
```
Comment (b-gent) [inline nrf_security/doc/backend_config.rst:564]: add secp224r1

File: nrf_security/doc/backend_config.rst [hunk 17]
```diff
@@ -372,8 +604,57 @@ Elliptic Curve Digital Signature Algorithm (ECDSA) support can be enabled be con
    * ECDSA support depends upon `ECC Configurations`_ being enabled.
    * The :ref:`nrf_cc3xx_mbedcrypto_readme` does not integrate on ECP layer.
      Only the top-level APIs for ECDSA are replaced.
-   * The :ref:`nrf_oberon_readme` only supports ECC curve secp256r1.
 
+Feature support
+===============
+
++-----------+-------------------+----------+-------------------+
+| Algorithm | Backend           | Key size | Curve type        |
```
Comment (b-gent) [inline nrf_security/doc/backend_config.rst:612]: remove column key size

File: nrf_security/doc/backend_config.rst [hunk 18]
```diff
@@ -372,8 +604,57 @@ Elliptic Curve Digital Signature Algorithm (ECDSA) support can be enabled be con
...
+|           |                   |          |                   |
+|           |                   |          |                   |
+|           |                   |          | Koblitz group:    |
+|           |                   |          | secp160k1         |
+|           |                   |          | secp192k1         |
+|           |                   |          | secp224k1         |
+|           |                   |          | secp256k1         |
+|           +-------------------+----------+-------------------+
+|           | nrf_oberon        | TBD      | NIST group:       |
+|           |                   |          | secp256r1         |
```
Comment (b-gent) [inline nrf_security/doc/backend_config.rst:631]: secp224r1

**Comment by b-gent** [comment]
ECJPAKE -> all backends support using secp256r1 , add this

**Comment by b-gent** [comment]
RSA - only cc3xx and mbed tls, limitation of keys to 2048 for cc3x, no limitation in mbed tls

**Comment by b-gent** [comment]
incorrect notes for SHA - oberon now supports sha-224

**Comment by b-gent** [comment]
ECC curves configurations - add columns for backends to the table and add check marks

**Comment by b-gent** [comment]
Version after recent changes:

[Backend configurations and supported features_3.pdf](https://github.com/nrfconnect/sdk-nrfxlib/files/5965814/Backend.configurations.and.supported.features_3.pdf)

**Comment by b-gent** [comment]
> ECC curves configurations - add columns for backends to the table and add check marks

I suggest to split this from PR and postpone a bit.

**Comment by b-gent** [comment]
PDF in landscape orientation for easier review:
[Uploading Backend configurations and supported features_3_landscape.pdf…]()

**Comment by b-gent** [comment]
Next (final?) iteration -> 
[Backend configurations and supported features_4.pdf](https://github.com/nrfconnect/sdk-nrfxlib/files/5984645/Backend.configurations.and.supported.features_4.pdf)

@frkv please check the changes we discussed yesterday:
- removing secp160r1/r1/k1
- Removing note that SHA1 is not supported by oberon
- add config parameters for oberon in chacha, poly, and chacha-poly tables

**Review by frkv — APPROVED** [review]
LGTM

**Review by tejlmand — APPROVED** [review]

File: nrf_security/doc/backend_config.rst [hunk 19]
```diff
@@ -336,8 +518,63 @@ ECC core support can be enabled by setting the :option:`CONFIG_MBEDTLS_ECP_C` Kc
...
+|           |                   +-------------+------------+
+|           |                   | Curve25519  | Curve25519 |
+|           +-------------------+-------------+------------+
+|           | nrf_oberon        | NIST        | secp256r1  |
+|           |                   |             +------------+
+|           |                   |             | secp224r1  |
+|           |                   +-------------+------------+
+|           |                   | Curve25519  | Curve25519 |
+|           |                   |             +------------+
+|           |                   |             | Ed25519    |
```
Comment (b-gent) [inline nrf_security/doc/backend_config.rst:552]: @frkv

Comment (frkv) [reply nrf_security/doc/backend_config.rst:552]: Ah yes.  Needs to be corrected. Is only available in PRs in mbed TLS at this moment

**Pull request merged**